### PR TITLE
FIX-2167: tempodb integer divide by zero error

### LINKED ISSUE #2167: tempodb integer divide by zero error
I'm just letting you know about potential problem. I did some experiments and somehow I broke tempo storage connected with minio (without replication). I have no idea how to reproduce it, but **from stack trace it seems like it needs to check for 0 value before dividing it somewhere**. I'm obviously using docker-compose setup and after I removed volume and started everything over again the issue is gone.

```
stack-tempo-1    | panic: runtime error: integer divide by zero
stack-tempo-1    | 
stack-tempo-1    | goroutine 632 [running]:
stack-tempo-1    | github.com/grafana/tempo/tempodb.(*timeWindowBlockSelector).windowForTime(...)
stack-tempo-1    |      /drone/src/tempodb/compaction_block_selector.go:182
stack-tempo-1    | github.com/grafana/tempo/tempodb.newTimeWindowBlockSelector({0x4000284cf0, 0x1, 0xd?}, 0x0, 0x0, 0x0, 0x2, 0x4)
stack-tempo-1    |      /drone/src/tempodb/compaction_block_selector.go:58 +0x604
stack-tempo-1    | github.com/grafana/tempo/tempodb.(*readerWriter).doCompaction(0x40003006c0)
stack-tempo-1    |      /drone/src/tempodb/compactor.go:92 +0xf4
stack-tempo-1    | github.com/grafana/tempo/tempodb.(*readerWriter).compactionLoop(0x40003006c0)
stack-tempo-1    |      /drone/src/tempodb/compactor.go:74 +0x64
stack-tempo-1    | created by github.com/grafana/tempo/tempodb.(*readerWriter).EnableCompaction
stack-tempo-1    |      /drone/src/tempodb/tempodb.go:407 +0x1d0
stack-tempo-1 exited with code 2
```

### PR DESCRIPTION
**Pull request by mghildiy**
**What this PR does**:
This PR solves the 'divison by zero' error occurring during compaction.

**Which issue(s) this PR fixes**:
Fixes #2167

**Checklist**
- [ config_test, tempodb_test] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@
 /cmd/tempo-serverless/vendor/
 /pkg/traceql/y.output
 private-key.key
+/vendor
```
Comment (joe-elliott) [inline .gitignore:16]: we need to leave the vendor folder in git

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ## main / unreleased
 
+* [BUGFIX] tempodb integer divide by zero error [#2167](https://github.com/grafana/tempo/issues/2167)
```
Comment (joe-elliott) [inline CHANGELOG.md:3]: please merge main or rebase. there is a new version in the changelog

File: tempodb/__debug_bin.exe
Comment (joe-elliott) [inline tempodb/__debug_bin.exe:1]: this should not be here

File: tempodb/compaction_block_selector.go [hunk 3]
```diff
@@ -54,6 +56,11 @@ func newTimeWindowBlockSelector(blocklist []*backend.BlockMeta, maxCompactionRan
 		MaxBlockBytes:        maxBlockBytes,
 	}
 
+	if maxCompactionRange == 0 {
```
Comment (joe-elliott) [inline tempodb/compaction_block_selector.go:59]: i believe this returns an invalid block selector which would make the compactor behave unpredictably.

perhaps we should make a `validateConfig` function like this:

https://github.com/grafana/tempo/blob/main/tempodb/config.go#L119

and then call it in `EnableCompaction`:

https://github.com/grafana/tempo/blob/main/tempodb/tempodb.go#L379

File: vendor/golang.org/x/tools/internal/gcimporter/gcimporter.go
Comment (joe-elliott) [inline vendor/golang.org/x/tools/internal/gcimporter/gcimporter.go:1]: we really shouldn't need any vendor changes for this fix

**Review by joe-elliott — COMMENTED** [review]
thanks for submitting a fix for this! i think you need to rebase on top of main or merge main to get up to date. that should resolve your go.mod issues.

**Comment by mghildiy** [comment]
I merged main into mny feature branch, and it somehow stops git working. For every git command I now get:

**fatal: not a git repository (or any of the parent directories): .git**

Never faced it with git. Anone having any idea?

File: tempodb/compaction_block_selector.go (hunk 3 above)
Comment (mghildiy) [reply tempodb/compaction_block_selector.go:59]: So I need to check that compactor.compaction.compaction_window is not zero. If it is, I log it as an error. But as I would just log, and not temrinate the program, that divison code would still execute, isn't?

**Comment by mghildiy** [comment]
I have made the recommended changes. Pleade have a look.

- A new validation method for CompactorConfig in config.go to check that compaction range is not 0
- Corresponding test in config_test.go.
- Call this new method in EnableCompaction method in tempodb.go, and if it returns error, just return from the method without doing anything more. 
- Corresponding test in tempodb_test.go

File: tempodb/tempodb.go [hunk 4]
```diff
@@ -377,6 +377,13 @@ func (rw *readerWriter) Shutdown() {
 
 // EnableCompaction activates the compaction/retention loops
 func (rw *readerWriter) EnableCompaction(ctx context.Context, cfg *CompactorConfig, c CompactorSharder, overrides CompactorOverrides) {
+	// If compactor configuration is not as expected, no need to go any further
+	err := cfg.validateCompactorConfig()
+	if err != nil {
+		level.Warn(log.Logger).Log(err.Error())
```
Comment (joe-elliott) [inline tempodb/tempodb.go:383]: nit: log at the error level?

File: tempodb/config.go [hunk 5]
```diff
@@ -116,6 +116,14 @@ type CompactorConfig struct {
 	CompactionCycle         time.Duration `yaml:"compaction_cycle"`
 }
 
+func (compactorConfig CompactorConfig) validateCompactorConfig() error {
```
Comment (joe-elliott) [inline tempodb/config.go:119]: nit: just name it "validate()". "CompactorConfig" is implied by the struct.

**Review by joe-elliott — COMMENTED** [review]
nice, thanks for restructuring. This PR is looking a lot cleaner now. Please clean up the changelog conflict and consider the small comments I've made and then we can merge.

**Comment by mghildiy** [comment]
Hmmm....I am a bit surprised as I saw a test failure. And when I ran that test(TestInstance_cantWriteToWAL) locally, there too it failed. But now PR is approved.

**Pull request merged**